clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -1614,14 +1614,40 @@ Decl *TemplateDeclInstantiator::VisitVarDecl(VarDecl *D,
   // Only add this if we aren't instantiating a variable template.  We'll end up
   // adding the VarTemplateSpecializationDecl later.
   if (!InstantiatingVarTemplate) {
-    SemaRef.addSyclVarDecl(Var);
+    if (SemaRef.getLangOpts().SYCLIsDevice &&
+        SemaRef.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(
+            Var->getType())) {
+      if (!Var->hasGlobalStorage() || Var->isLocalVarDeclOrParm()) {
+        SemaRef.Diag(D->getLocation(),
+                     diag::err_sycl_device_global_incorrect_scope);
+        return nullptr;
+      }
+
+      if (Var->getAccess() == AS_private)
+        SemaRef.Diag(D->getLocation(),
+                     diag::err_sycl_device_global_not_publicly_accessible)
+            << Var;
```

Comment:
Should a `return nullptr;` be under this `if`?
It seems other cases added by this patch in this file have it.

Reply:
It seems I added the nullptr needlessly in the process of debugging something else. It seems better to simply get rid of it.

Reply:
I actually didn't mind having an early exit like this. In fact, the function that is being modified here has this kind of exit on line 1551 of this file. Also, double diagnosing that appeared in the test due to its removal might be concerning.
@elizabethandrews , WDYT?

Reply:
From a quick look it does seem like early exist makes sense here. I don't understand why removing the `nullptr` after L1222 fixed the diagnostic in L65 of the test actually. Can you explain @schittir ?

Reply:
When `nullptr` is returned after the first VarDecl (L63 of test), the control doesn't come back to the second VarDecl (L65) in the instantiation.  In the call stack of `Instantiator.TransformStmt(S)` (L3605 of SemaTemplateInstantiateDecl.cpp) is a call to `Decl *Transformed = getDerived().TransformDefinition(D->getLocation(), D);` (L7737 of TreeTransform.h), which returns empty.

Reply:
That makes sense, thanks @schittir for the explanation.

Reply:
Oh I see. So L7738 above will be true after the first VarDecl and the rest aren't processed. I wonder why they did it that way...Thanks for the explanation @schittir!